amazonsns/publisher.go

```diff
@@ -0,0 +1,66 @@
+package amazonsns
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sns"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+
+	watermill_amazonsqs "github.com/ThreeDotsLabs/watermill-amazonsqs"
+	"github.com/ThreeDotsLabs/watermill-amazonsqs/amazonsqs"
+)
+
+type Publisher struct {
+	config watermill_amazonsqs.PublisherConfig
+	logger watermill.LoggerAdapter
+	sns    *sns.SNS
+}
+
+func NewPublisher(config watermill_amazonsqs.PublisherConfig, logger watermill.LoggerAdapter) (*Publisher, error) {
+	config.AWSConfig = amazonsqs.SetEndPoint(config.AWSConfig)
+	sess, err := session.NewSession(&config.AWSConfig)
+	if err != nil {
+		// TODO wrap
+		return nil, err
+	}
+
+	return &Publisher{
+		sns:    sns.New(sess),
+		config: config,
+		logger: logger,
+	}, nil
+}
+
+func (p Publisher) Publish(topic string, messages ...*message.Message) error {
+	// TODO method for generating
+	topicInfo, err := p.sns.CreateTopic(&sns.CreateTopicInput{
+		Name: aws.String(topic),
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, msg := range messages {
+		sqsMsg, err := p.config.Marshaler.Marshal(msg)
+		if err != nil {
+			return err
+		}
+
+		p.logger.Info("Sending message", watermill.LogFields{"msg": msg})
+		_, err = p.sns.Publish(&sns.PublishInput{
+			TopicArn: topicInfo.TopicArn,
+			Message:  sqsMsg.Body,
+		})
+		if err != nil {
```

Comment:
are you sure you want to stop here? I'm thinking about scenario where I have 2 messages. The first one succeeds but the second not. Should I resend both of them? How can I know which messages were sent or not? :P

Reply:
SNS is a simple delivery system to multiple queues/topics and http endpoints the reader of the messages that are located in persistent queues will be responsable to either Ack() or not - if the message is not Acked then AWS will try delivering the message again until success.  Is that what you meant?

Reply:
This is a valid concern and it's not unique to this implementation. If you're publishing more than one message in batch it's possible that the process will be interrupted in the middle and we can't really do anything about it.

Kafka implementation is similar: https://github.com/ThreeDotsLabs/watermill-kafka/blob/master/pkg/kafka/publisher.go#L91

If you need more control, you need to call `Publish` one-by-one with each message.